src/arch/x86_64/multiboot.rs

```diff
@@ -92,78 +92,63 @@ impl DeviceTree {
 }
 
 pub fn find_kernel() -> &'static [u8] {
-	use core::cmp;
-
 	paging::clean_up();
 	// Identity-map the Multiboot information.
 	unsafe {
 		assert!(mb_info > 0, "Could not find Multiboot information");
 		info!("Found Multiboot information at {:#x}", mb_info);
 	}
-	let page_address = unsafe { mb_info.align_down(Size4KiB::SIZE as usize) };
-	paging::map::<Size4KiB>(page_address, page_address, 1, PageTableFlags::empty());
+	unsafe { paging::map::<Size4KiB>(mb_info, mb_info, 1, PageTableFlags::empty()) };
 
 	let mut mem = Mem;
 	// Load the Multiboot information and identity-map the modules information.
 	let multiboot = unsafe { Multiboot::from_ptr(mb_info as u64, &mut mem).unwrap() };
-	let modules_address = multiboot
-		.modules()
-		.expect("Could not find a memory map in the Multiboot information")
-		.next()
-		.expect("Could not find first map address")
-		.start as usize;
-	let page_address = modules_address.align_down(Size4KiB::SIZE as usize);
-	paging::map::<Size4KiB>(page_address, page_address, 1, PageTableFlags::empty());
 
 	// Iterate through all modules.
 	// Collect the start address of the first module and the highest end address of all modules.
-	let modules = multiboot.modules().unwrap();
-	let mut found_module = false;
-	let mut start_address = 0;
-	let mut end_address = 0;
-
-	for m in modules {
-		found_module = true;
-
-		if start_address == 0 {
```

Comment:
I assumed this line is to check if this is the first iteration of the loop (because the variable is initialized to 0) but if it is possible that some of the modules have 0 as their start value, my assumption is incorrect and the new code behaves differently. I don't think that is possible, though.